src/main/java/updater/FileDownloader.java

```diff
@@ -0,0 +1,161 @@
+package updater;
+
+import javafx.application.Platform;
+import org.apache.commons.io.IOUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import ui.UI;
+import util.DialogMessage;
+import util.events.Event;
+import util.events.UpdateManagerProgressWindowEvent;
+
+import java.io.*;
+import java.net.URI;
+import java.net.URLConnection;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * A class in charge of downloading file.
+ * It has a progress tracker which can be used to show progress window.
+ */
+public class FileDownloader {
+    private static final Logger logger = LogManager.getLogger(FileDownloader.class.getName());
+
+    public static final int CONNECTION_TIMEOUT = 15000;
+    public static final int READ_CONNECTION_TIMEOUT = 30000;
+
+    private long totalFileSizeInBytes;
+
+    private final ProgressTracker progressTracker;
+
+    public FileDownloader() {
+        progressTracker = new ProgressTracker();
+    }
+
+    /**
+     * Set if current download can have its progress shown
+     *
+     * @param enableProgressWindow
+     */
+    public void setEnableProgressWindow(boolean enableProgressWindow) {
+        UI.events.triggerEvent(new UpdateManagerProgressWindowEvent(enableProgressWindow));
+    }
+
+    /**
+     * Set current download's progress window title and message
+     *
+     * @param windowTitle
+     * @param windowMessage
+     */
+    public void setProgressTrackerWindowDescription(String windowTitle, String windowMessage) {
+        this.progressTracker.setProgressWindowDetails(windowTitle, windowMessage);
+    }
+
+    /**
+     * Download file from a URL to a local destination file
+     *
+     * @param source URI of file source, expect to be a valid URL
+     * @param dest local destination file
+     */
+    public void downloadFileFromUrlToFile(URI source, File dest) {
```

Comment:
Can this class download multiple files simultaneously, or be used multiple times to download a file? If not, maybe these parameters should be taken in the constructor, so each FileDownloader maps to one download, one file, and (eventually) one progress bar. Then this method could take no parameters and just start the download.